Netbox upgrade to 4.1.1 broke api schema

### Validity Version

3.0.1

### NetBox Version

4.1.1

### Python Version

3.12.3

### Steps to Reproduce

https://netbox.absgroup.ru/api/schema/?format=json started to traceback after upgrade netbox to 4.1.1.


### Traceback

```shell
/opt/netbox/venv/lib/python3.12/site-packages/validity/api/serializers.py: Warning [SerializedStateView > SerializedStateSerializer]: unable to resolve type hint for function "get_count". Consider using a type hint or @extend_schema_field. Defaulting to string.

/opt/netbox/venv/lib/python3.12/site-packages/validity/api/serializers.py: Warning [SerializedStateView > SerializedStateSerializer > SerializedStateItemSerializer]: unable to resolve type hint for function "get_serialized". Consider using a type hint or @extend_schema_field. Defaulting to string.

/opt/netbox/netbox/extras/api/serializers_/customfields.py: Warning [CustomFieldChoiceSetViewSet > CustomFieldChoiceSetSerializer]: unable to resolve type hint for function "choices_count". Consider using a type hint or @extend_schema_field. Defaulting to string.

/opt/netbox/netbox/extras/api/serializers_/customfields.py: Warning [CustomFieldViewSet > CustomFieldSerializer > CustomFieldChoiceSetSerializer]: unable to resolve type hint for function "choices_count". Consider using a type hint or @extend_schema_field. Defaulting to string.

/opt/netbox/netbox/extras/api/serializers_/customfields.py: Warning [CustomFieldViewSet > WritableCustomFieldSerializer > CustomFieldChoiceSetSerializer]: unable to resolve type hint for function "choices_count". Consider using a type hint or @extend_schema_field. Defaulting to string.

/opt/netbox/netbox/ipam/filtersets.py: Warning [ServiceViewSet > ServiceFilterSet]: model field "IPAddressField" has no mapping in ModelSerializer. It may be a deprecated field. Defaulting to "string"

/opt/netbox/venv/lib/python3.12/site-packages/validity/api/serializers.py: Warning [NameSetViewSet > NameSetSerializer]: unable to resolve type hint for function "effective_definitions". Consider using a type hint or @extend_schema_field. Defaulting to string.

/opt/netbox/venv/lib/python3.12/site-packages/validity/api/serializers.py: Warning [ComplianceReportViewSet > ComplianceReportSerializer]: could not resolve field on model <class 'validity.models.report.ComplianceReport'> with path "device_count". This is likely a custom field that does some unknown magic. Maybe consider annotating the field/property? Defaulting to "string". (Exception: ComplianceReport has no field named 'device_count')

/opt/netbox/venv/lib/python3.12/site-packages/validity/api/serializers.py: Warning [ComplianceReportViewSet > ComplianceReportSerializer]: could not resolve field on model <class 'validity.models.report.ComplianceReport'> with path "test_count". This is likely a custom field that does some unknown magic. Maybe consider annotating the field/property? Defaulting to "string". (Exception: ComplianceReport has no field named 'test_count')

/opt/netbox/venv/lib/python3.12/site-packages/validity/api/serializers.py: Warning [ComplianceReportViewSet > ComplianceReportSerializer]: could not resolve field on model <class 'validity.models.report.ComplianceReport'> with path "total_passed". This is likely a custom field that does some unknown magic. Maybe consider annotating the field/property? Defaulting to "string". (Exception: ComplianceReport has no field named 'total_passed')

/opt/netbox/venv/lib/python3.12/site-packages/validity/api/serializers.py: Warning [ComplianceReportViewSet > ComplianceReportSerializer]: could not resolve field on model <class 'validity.models.report.ComplianceReport'> with path "total_count". This is likely a custom field that does some unknown magic. Maybe consider annotating the field/property? Defaulting to "string". (Exception: ComplianceReport has no field named 'total_count')

/opt/netbox/venv/lib/python3.12/site-packages/validity/api/serializers.py: Warning [ComplianceReportViewSet > ComplianceReportSerializer]: could not resolve field on model <class 'validity.models.report.ComplianceReport'> with path "low_passed". This is likely a custom field that does some unknown magic. Maybe consider annotating the field/property? Defaulting to "string". (Exception: ComplianceReport has no field named 'low_passed')

/opt/netbox/venv/lib/python3.12/site-packages/validity/api/serializers.py: Warning [ComplianceReportViewSet > ComplianceReportSerializer]: could not resolve field on model <class 'validity.models.report.ComplianceReport'> with path "low_count". This is likely a custom field that does some unknown magic. Maybe consider annotating the field/property? Defaulting to "string". (Exception: ComplianceReport has no field named 'low_count')

/opt/netbox/venv/lib/python3.12/site-packages/validity/api/serializers.py: Warning [ComplianceReportViewSet > ComplianceReportSerializer]: could not resolve field on model <class 'validity.models.report.ComplianceReport'> with path "middle_passed". This is likely a custom field that does some unknown magic. Maybe consider annotating the field/property? Defaulting to "string". (Exception: ComplianceReport has no field named 'middle_passed')

/opt/netbox/venv/lib/python3.12/site-packages/validity/api/serializers.py: Warning [ComplianceReportViewSet > ComplianceReportSerializer]: could not resolve field on model <class 'validity.models.report.ComplianceReport'> with path "middle_count". This is likely a custom field that does some unknown magic. Maybe consider annotating the field/property? Defaulting to "string". (Exception: ComplianceReport has no field named 'middle_count')

/opt/netbox/venv/lib/python3.12/site-packages/validity/api/serializers.py: Warning [ComplianceReportViewSet > ComplianceReportSerializer]: could not resolve field on model <class 'validity.models.report.ComplianceReport'> with path "high_passed". This is likely a custom field that does some unknown magic. Maybe consider annotating the field/property? Defaulting to "string". (Exception: ComplianceReport has no field named 'high_passed')

/opt/netbox/venv/lib/python3.12/site-packages/validity/api/serializers.py: Warning [ComplianceReportViewSet > ComplianceReportSerializer]: could not resolve field on model <class 'validity.models.report.ComplianceReport'> with path "high_count". This is likely a custom field that does some unknown magic. Maybe consider annotating the field/property? Defaulting to "string". (Exception: ComplianceReport has no field named 'high_count')

/opt/netbox/venv/lib/python3.12/site-packages/validity/api/serializers.py: Warning [ComplianceReportViewSet > ComplianceReportSerializer]: unable to resolve type hint for function "get_results_url". Consider using a type hint or @extend_schema_field. Defaulting to string.

Internal Server Error: /api/schema/

Traceback (most recent call last):

  File "/opt/netbox/venv/lib/python3.12/site-packages/django/core/handlers/exception.py", line 55, in inner

    response = get_response(request)

               ^^^^^^^^^^^^^^^^^^^^^

  File "/opt/netbox/venv/lib/python3.12/site-packages/django/core/handlers/base.py", line 197, in _get_response

    response = wrapped_callback(request, *callback_args, **callback_kwargs)

               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/opt/netbox/venv/lib/python3.12/site-packages/django/utils/decorators.py", line 188, in _view_wrapper

    result = _process_exception(request, e)

             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/opt/netbox/venv/lib/python3.12/site-packages/django/utils/decorators.py", line 186, in _view_wrapper

    response = view_func(request, *args, **kwargs)

               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/opt/netbox/venv/lib/python3.12/site-packages/django/views/decorators/csrf.py", line 65, in _view_wrapper

    return view_func(request, *args, **kwargs)

           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/opt/netbox/venv/lib/python3.12/site-packages/django/views/generic/base.py", line 104, in view

    return self.dispatch(request, *args, **kwargs)

           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/opt/netbox/venv/lib/python3.12/site-packages/rest_framework/views.py", line 509, in dispatch

    response = self.handle_exception(exc)

               ^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/opt/netbox/venv/lib/python3.12/site-packages/rest_framework/views.py", line 469, in handle_exception

    self.raise_uncaught_exception(exc)

  File "/opt/netbox/venv/lib/python3.12/site-packages/rest_framework/views.py", line 480, in raise_uncaught_exception

    raise exc

  File "/opt/netbox/venv/lib/python3.12/site-packages/rest_framework/views.py", line 506, in dispatch

    response = handler(request, *args, **kwargs)

               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/opt/netbox/venv/lib/python3.12/site-packages/drf_spectacular/views.py", line 84, in get

    return self._get_schema_response(request)

           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/opt/netbox/venv/lib/python3.12/site-packages/drf_spectacular/views.py", line 92, in _get_schema_response

    data=generator.get_schema(request=request, public=self.serve_public),

         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/opt/netbox/venv/lib/python3.12/site-packages/drf_spectacular/generators.py", line 285, in get_schema

    paths=self.parse(request, public),

          ^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/opt/netbox/venv/lib/python3.12/site-packages/drf_spectacular/generators.py", line 256, in parse

    operation = view.schema.get_operation(

                ^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/opt/netbox/venv/lib/python3.12/site-packages/drf_spectacular/openapi.py", line 112, in get_operation

    operation['responses'] = self._get_response_bodies()

                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/opt/netbox/venv/lib/python3.12/site-packages/drf_spectacular/openapi.py", line 1374, in _get_response_bodies

    return {'200': self._get_response_for_code(response_serializers, '200', direction=direction)}

                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/opt/netbox/venv/lib/python3.12/site-packages/drf_spectacular/openapi.py", line 1430, in _get_response_for_code

    component = self.resolve_serializer(serializer, direction)

                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/opt/netbox/venv/lib/python3.12/site-packages/drf_spectacular/openapi.py", line 1614, in resolve_serializer

    component.schema = self._map_serializer(serializer, direction, bypass_extensions)

                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/opt/netbox/venv/lib/python3.12/site-packages/drf_spectacular/openapi.py", line 930, in _map_serializer

    schema = self._map_basic_serializer(serializer, direction)

             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/opt/netbox/venv/lib/python3.12/site-packages/drf_spectacular/openapi.py", line 1029, in _map_basic_serializer

    schema = self._map_serializer_field(field, direction)

             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/opt/netbox/venv/lib/python3.12/site-packages/drf_spectacular/openapi.py", line 680, in _map_serializer_field

    schema = self._map_serializer_field(field.child_relation, direction)

             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/opt/netbox/venv/lib/python3.12/site-packages/drf_spectacular/openapi.py", line 653, in _map_serializer_field

    schema = serializer_field_extension.map_serializer_field(self, direction)

             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/opt/netbox/netbox/core/api/schema.py", line 272, in map_serializer_field

    component = auto_schema.resolve_serializer(self.target.serializer, direction)

                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

  File "/opt/netbox/venv/lib/python3.12/site-packages/drf_spectacular/openapi.py", line 1601, in resolve_serializer

    assert_basic_serializer(serializer)

  File "/opt/netbox/venv/lib/python3.12/site-packages/drf_spectacular/plumbing.py", line 186, in assert_basic_serializer

    assert is_basic_serializer(serializer), (

           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

AssertionError: internal assumption violated because we expected a basic serializer here and instead got a "functools.partial(<class 'tenancy.api.serializers_.tenants.TenantSerializer'>, nested=True)". This may be the result of another app doing some unexpected magic or an invalid internal call. Feel free to report this as a bug at https://github.com/tfranzel/drf-spectacular/issues
```

--- a/validity/api/serializers.py
+++ b/validity/api/serializers.py
@@ -165,7 +165,7 @@ class ComplianceReportSerializer(NetBoxModelSerializer):
         )
         brief_fields = ("id", "url", "display")
 
-    def get_results_url(self, obj):
+    def get_results_url(self, obj) -> str:
         results_url = reverse("plugins-api:validity-api:compliancetestresult-list", request=self.context["request"])
         return results_url + f"?report_id={obj.pk}"
 

--- a/validity/netbox_changes/__init__.py
+++ b/validity/netbox_changes/__init__.py
@@ -6,6 +6,7 @@ different versions of NetBox together
 from functools import partial
 
 from validity import config
+from validity.api.helpers import nested_factory
 
 
 if config.netbox_version >= "4.1.0":
@@ -25,4 +26,4 @@ if config.netbox_version < "4.0.0":
 else:
     from tenancy.api.serializers import TenantSerializer as __TenantSerializer
 
-    NestedTenantSerializer = partial(__TenantSerializer, nested=True)
+    NestedTenantSerializer = nested_factory(__TenantSerializer, nb_version=config.netbox_version)

--- a/validity/tests/test_api.py
+++ b/validity/tests/test_api.py
@@ -222,3 +222,8 @@ def test_run_tests(admin_client, di, post_body, status_code):
         assert resp.status_code == status_code
         if resp.status_code == HTTPStatus.OK:
             launcher.assert_called_once()
+
+
+def test_api_schema_is_not_broken(admin_client):
+    resp = admin_client.get("/api/schema/")
+    assert resp.status_code == HTTPStatus.OK

